lib/grape/middleware/base.rb

```diff
@@ -27,7 +31,20 @@ def call!(env)
         ensure
           after_response = after
         end
-        after_response || @app_response
+
+        response = after_response || @app_response
+        reflect_headers response
+        response
+      end
+
+      def reflect_headers(response)
+        return unless headers.is_a?(Hash)
+        response_headers =
+          case response
+          when Rack::Response then response.headers
+          when Array          then response[1]
+          end
+        response_headers.merge!(headers)
```

Comment:
If the API developer made a mistake and `response` is neither, we get a `nil`, I would stay defensive and do `merge!() ... if response_headers`.